### PR TITLE
Add IndexOf function to ItemsSourceView.

### LINKED ISSUE #1828: ItemsSourceView should expose IndexOf
There are cases where we need IndexOf to be exposed in ItemsSourceView. If the ItemsSource supports this we can provide a value. This is currently needed for NavigationView and RadioButtons and we do have it internally. We just need to expose it via ItemsSourceView.

### PR DESCRIPTION
**Pull request by chingucoding**
## Description

This PR adds the IndexOf function.

API spec (in progress): https://github.com/microsoft/microsoft-ui-xaml-specs/pull/74
## Motivation and Context

Closes #1828 
## How Has This Been Tested?

Not tested so far, how should this be tested?

### CODE REVIEW

File: dev/Repeater/ItemsRepeater.idl [hunk 1]
```diff
@@ -106,6 +106,9 @@ unsealed runtimeclass ItemsSourceView: Windows.UI.Xaml.Interop.INotifyCollection
     Boolean HasKeyIndexMapping{ get; };
     String KeyFromIndex(Int32 index);
     Int32 IndexFromKey(String key);
+
+    [WUXC_VERSION_PREVIEW]
+    Int32 IndexOf(IInspectable item);
```
Comment (ranjeshj) [inline dev/Repeater/ItemsRepeater.idl]: For Testing, you can create an api test in [https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Repeater/APITests/InspectingDataSourceTests.cs](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Repeater/APITests/InspectingDataSourceTests.cs) and just call IndexOf with different values.

Comment (ranjeshj) [inline dev/Repeater/ItemsRepeater.idl]: >IndexOf [](start = 10, length = 7)

I think we internally cast the ItemsSourceView to InspectingDataSource to get access to the IndexOf method. Can you do a search and remove those? now that this is public, we can just use it without casting.

Comment (chingucoding) [reply dev/Repeater/ItemsRepeater.idl]: Good idea, added a new test there.

Comment (chingucoding) [reply dev/Repeater/ItemsRepeater.idl]: I've switched the areas in NavigationView that used that casting pattern to use IndexOf to just use the new method. However in the RadioButtons case, it turns out, that uppon initial loading, they crash when using the ItemsSourceView method.

Would you take a look if the way I implemented the IndexOf function is the reason for that?
From debugging, it seems that calling the virtual method ends up being the problem here.

Comment (ranjeshj) [reply dev/Repeater/ItemsRepeater.idl]: I'm sorry - I think we will have to wait on removing the casts until the API is in stable release, otherwise in stable releases the API still won't be available.

Comment (chingucoding) [reply dev/Repeater/ItemsRepeater.idl]: Should I revert my latest commit, and we add that function as preview now or should we wait until the spec has gone through review?

Comment (ranjeshj) [reply dev/Repeater/ItemsRepeater.idl]: Yes. I think that is the best option for now. Thanks.

Comment (chingucoding) [reply dev/Repeater/ItemsRepeater.idl]: Done!

File: dev/Repeater/ItemsSourceView.h [hunk 2]
```diff
@@ -22,6 +22,8 @@ class ItemsSourceView :
     void CollectionChanged(winrt::event_token const& token);
 #pragma endregion
 
+    int IndexOf(winrt::IInspectable const& value);
```
Comment (ranjeshj) [inline dev/Repeater/ItemsSourceView.h]: This can move into the region above and while we are here can we change it from IDataSource to ItemsSourceView in the pragma.

File: dev/Repeater/InspectingDataSource.h [hunk 3]
```diff
@@ -24,6 +24,7 @@ class InspectingDataSource :
 #pragma endregion
 
     int IndexOf(winrt::IInspectable const& value);
+    int IndexOfCore(winrt::IInspectable const& value);
```
Comment (ranjeshj) [inline dev/Repeater/InspectingDataSource.h]: move into region above and rename IdataSourceOverrides into ItemsSourceViewOverrides

File: dev/Repeater/InspectingDataSource.h [hunk 4]
```diff
@@ -24,6 +24,7 @@ class InspectingDataSource :
 #pragma endregion
 
     int IndexOf(winrt::IInspectable const& value);
```
Comment (ranjeshj) [inline dev/Repeater/InspectingDataSource.h]: I think this is not used anymore.

File: dev/Repeater/APITests/InspectingDataSourceTests.cs [hunk 5]
```diff
@@ -115,6 +115,22 @@ public void VerifyUniqueIdMappingInterface()
             });
         }
 
+        [TestMethod]
+        public void VerifyIndexOfBehavior()
+        {
+            RunOnUIThread.Execute(() =>
+            {
+                var data = new ObservableVectorWithUniqueIds(Enumerable.Range(0, 10));
```
Comment (ranjeshj) [inline dev/Repeater/APITests/InspectingDataSourceTests.cs]: can you please also add tests for Enumerable and  ObservableCollection

File: dev/Repeater/InspectingDataSource.h (hunk 4 above)
Comment (chingucoding) [reply dev/Repeater/InspectingDataSource.h]: IndexOf is still being used as we currently rely on casting unfortunately until the API is public.

Comment (ranjeshj) [reply dev/Repeater/InspectingDataSource.h]: Ah right :) can you please add a comment that it can be removed once the API is public ?

File: dev/Repeater/ItemsRepeater.idl (hunk 1 above)
Comment (ranjeshj) [reply dev/Repeater/ItemsRepeater.idl]: Yes. Thats a good idea.

File: dev/Repeater/APITests/InspectingDataSourceTests.cs (hunk 5 above)
Comment (chingucoding) [reply dev/Repeater/APITests/InspectingDataSourceTests.cs]: Added them. Using just an IEnumerable, calling IndexOf throws an exception, that is expected right (as casting to IVector does not work)?

Comment (ranjeshj) [reply dev/Repeater/APITests/InspectingDataSourceTests.cs]: Hmm.. weird, can you debug that case? I think [IEnumerable maps to IIterable](https://docs.microsoft.com/en-us/uwp/api/Windows.Foundation.Collections.IIterable_T_), so it should go through WrapIterable call which will end up creating a vector and that should work...

Comment (chingucoding) [reply dev/Repeater/APITests/InspectingDataSourceTests.cs]: It seems that Enumerable.Range() is the problem, as it generates an IEnumerable<int>. That seems to be problematic as int will be passed by value, not by reference.
Is this something we need to take care of in the API (by allowing ref's for example) ?

**Comment by chingucoding** [comment]
During API review @oldnewthing asked about passing `null` to IndexOf. We came to the conclusion that passing null should be valid. However the current implementation simply returns -1 if we pass `null`, regardless whether `null` is in the collection or not. Should I update this PR to also handle null as outlined in the API review or was there a specific reason to treat null different here?

**Comment by ranjeshj** [comment]
> During API review @oldnewthing asked about passing `null` to IndexOf. We came to the conclusion that passing null should be valid. However the current implementation simply returns -1 if we pass `null`, regardless whether `null` is in the collection or not. Should I update this PR to also handle null as outlined in the API review or was there a specific reason to treat null different here?

Let's wait until the api review PR is closed before making this update in case we decide otherwise.

**Comment by ranjeshj** [comment]
> > During API review @oldnewthing asked about passing `null` to IndexOf. We came to the conclusion that passing null should be valid. However the current implementation simply returns -1 if we pass `null`, regardless whether `null` is in the collection or not. Should I update this PR to also handle null as outlined in the API review or was there a specific reason to treat null different here?
> 
> Let's wait until the api review PR is closed before making this update in case we decide otherwise.

@chingucoding Let's handle the null case now that we have gone through the review. Thanks!

**Comment by chingucoding** [comment]
Thanks for reminding, add test and support for null values now.
